cadctl/cmd/cluster-missing/cluster-missing.go

```diff
@@ -143,39 +142,194 @@ func evaluateMissingCluster(awsClient aws.Client, ocmClient ocm.Client, pdClient
+		}
+		fmt.Println("Primary has been alerted")
+		return fmt.Errorf("Failed to remove CCAM Limited Support reason from cluster %s: %w", externalClusterID, originalErr)
+	}
+	if removedReasons{
+		fmt.Println("Removed CCAM Limited Support reason from cluster")
+	} else {
+		fmt.Println("No CCAM Limited Support reasons needed to be removed")
+	}
+
+	// Remove 'Cluster Has Gone Missing' Limited Support reasons added by CAD
+	fmt.Println("Removing 'Cluster Has Gone Missing' Limited Support reasons")
+	retries = 1
+	removedReasons, err = chgmClient.RemoveCHGMLimitedSupport(externalClusterID)
+	for err != nil && retries <= 3 {
+		fmt.Printf("Failed to remove CHGM Limited Support reason from cluster %s: %v\n", externalClusterID, err)
+
+		// Sleep for a time, based on the number of retries (up to 300 seconds)
+		sleepBeforeRetrying(retries, 300)
+		retries++
+
+		removedReasons, err = chgmClient.RemoveCHGMLimitedSupport(externalClusterID)
+	}
+	// After 3 retries, if the LS reason hasn't been removed - alert Primary
+	if err != nil {
+		fmt.Println("Failed 3 times to remove CHGM Limited Support reason from cluster. Attempting to alert Primary.")
+		originalErr := err
+		retries = 1
+
+		err = chgmClient.CreateIncidentForLimitedSupportRemovalFailure(originalErr, externalClusterID, serviceID)
+		for err != nil {
+			fmt.Printf("Failed to alert Primary (attempt %d). Err: %v\n", retries, err)
+			fmt.Println("Retrying")
+
+			sleepBeforeRetrying(retries, 300)
+			retries++
+
+			err = chgmClient.CreateIncidentForLimitedSupportRemovalFailure(originalErr, externalClusterID, serviceID)
+
+		}
+		fmt.Println("Primary has been alerted")
+		return fmt.Errorf("Failed to remove CHGM Limited Support reason from cluster %s: %w", externalClusterID, originalErr)
+	}
+	if removedReasons {
+		fmt.Println("Removed CHGM Limited Support reason from cluster")
+	} else {
+		fmt.Println("No CHGM Limited Support reasons needed to be removed")
+	}
 	return nil
 }
 
+// retryUntilServiceObtained attempts to retrieve the serviceID from the PD payload, retrying until successful
+func retryUntilServiceObtained(chgmClient chgm.Client) string {
+		// Retrieve the service to alert on, retrying until we succeed
+		serviceID, err := chgmClient.ExtractServiceIDFromPayload(payloadPath, pagerduty.RealFileReader{})
+		retries := 1
+		for err != nil {
+			fmt.Printf("Failed to retrieve service from PagerDuty, retrying (attempt %d). Error encountered: %v\n", retries, err)
+
+			// Sleep for a time, based on the number of retries (up to 300 seconds)
+			sleepBeforeRetrying(retries, 300)
+			retries++
+
+			serviceID, err = chgmClient.ExtractServiceIDFromPayload(payloadPath, pagerduty.RealFileReader{})
+		}
+		return serviceID
+}
+
+// sleepBeforeRetrying sleeps with an exponential backoff based on the current retry, up to the given maximum sleep time.
+func sleepBeforeRetrying(currentRetry, maxSleepSeconds int) {
```

Comment:
Do we want an ability to cap the number of retries with this function directly?

I noticed that we sometimes manually cap `retries <= 3` when this is called and then sometimes forget to manually cap it. I think if we add an input for `maxRetries` and also return an `error`, we can return a non-nil error if the amount of retries has exceeded and we can just rely on the infinite for loop to handle `for err != nil`

Reply:
Retry logic is only present when evaluating restored clusters (cluster's that had their CHGM alert resolve), and the number of retries is only capped when attempting to remove clusters from Limited Support after a successful investigation. This was intentional, since, if we fail to remove Limited Support reasons after several attempts, this likely indicates a non-transient failure that needs to be escalated to Primary (ie- OCM outage, malformed reason, CAD misconfiguration). 

The number of retries is left unbounded in all cases where we need to escalate a condition to Primary. The reasoning here being, unlike cases where a CHGM has just triggered (evaluating a missing cluster), there is no other fallback option. If the investigation fails when the CHGM triggers, the alert will eventually escalate to Primary for examination. But if the investigation fails when the CHGM resolves, there is no signal that will escalate to Primary by default,  and we risk leaving clusters in Limited Support beyond the intended timeframe.

Reply:
Since CAD doesn't follow the typical operator reconcile-loop architecture, I don't think this is as much of a concern (we don't run the risk of blocking other cluster evaluations by suspending the process for X amount of time, etc).